// confinode/lib/Confinode/ConfinodeOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const os_1 = require("os");
const messages_1 = require("../messages");
// eslint-disable-next-line @typescript-eslint/no-empty-function
function noop() { }
function defaultLogger(message) {
    ;
    (message.level === messages_1.Level.Error
        ? // eslint-disable-next-line no-console
            console.error
        : message.level === messages_1.Level.Warning
            ? // eslint-disable-next-line no-console
                console.log
            : noop)(message.toString());
}
/**
 * Type guard to check if file names are actually filters rather than real file descriptions.
 *
 * @param files - The variable to check.
 * @returns True if files are actually a filter array.
 */
function filesAreFilters(files) {
    return files.length === 0 || typeof files[0] === 'function';
}
exports.filesAreFilters = filesAreFilters;
/**
 * Some default configuration options.
 */
exports.defaultConfig = {
    cache: true,
    searchStop: os_1.homedir(),
    logger: defaultLogger,
    customLoaders: {},
};
